File: MatlabOctaveCodeSnippets/snip_signals_fundamental_frequency.m
Fs=44100; %sampling frequency
Ts=1/Fs; %sampling interval
minF0Frequency=80; %minimum F0 frequency in Hz
maxF0Frequency=300; %minimum F0 frequency in Hz
minF0Period = 1/minF0Frequency; %correponding F0 (sec)
maxF0Period = 1/maxF0Frequency; %correponding F0 (sec)
Nbegin=round(maxF0Period/Ts);%number of lags for max freq.
Nend=round(minF0Period/Ts); %number of lags for min freq.
if 0 %record sound or test with 300 Hz cosine
    r = audiorecorder(Fs, 16, 1);%object audiorecorder
    disp('Started recording. Say a vowel a, e, i, o or u')
    recordblocking(r,2);%record 2 s and store in object r
    disp('finished recording');
    y=double(getaudiodata(r, 'int16'));%get recorded data
else %test with a cosine
    y=cos(2*pi*300*[0:2*Fs-1]*Ts); %300 Hz, duration 2 secs
end
subplot(211); plot(Ts*[0:length(y)-1],y);
xlabel('time (s)'); ylabel('Signal y(t)')
[R,lags]=xcorr(y,Nend,'biased'); %ACF with max lag Nend
subplot(212); %autocorrelation with normalized abscissa
plot(lags*Ts,R); xlabel('lag (s)');
ylabel('Autocorrelation of y(t)')
firstIndex = find(lags==Nbegin); %find index of lag
Rpartial = R(firstIndex:end); %just the region of interest
[Rmax, relative_index_max]=max(Rpartial);
%Rpartial was just part of R, so recalculate the index:
index_max = firstIndex - 1 + relative_index_max;
lag_max = lags(index_max); %get lag corresponding to index
hold on; %show the point:
plot(lag_max*Ts,Rmax,'xr','markersize',20);
F0 = 1/(lag_max*Ts); %estimated F0 frequency (Hz)
fprintf('Rmax=%g lag_max=%g T=%g (s) Freq.=%g Hz\n',...
    Rmax,lag_max,lag_max*Ts,F0);
t=0:Ts:2; soundsc(cos(2*pi*3*F0*t),Fs); %play freq. 3*F0

File: PythonCodeSnippets/snip_signals_fundamental_frequency.py
import matplotlib.pyplot as plt
import numpy as np
import pylab as pl
import scipy.signal as sp
from scipy.io.wavfile import write as wr

Fs = 44100  # sampling frequency
Ts = 1 / Fs  # sampling interval
minFOFrenquency = 80  # minimum F0 frequency in Hz
maxFOFrenquency = 300  # minimum F0 frequency in Hz
minFOPeriod = 1 / minFOFrenquency  # correponding F0 (sec)
maxFOPeriod = 1 / maxFOFrenquency  # correponding F0 (sec)
Nbegin = round(maxFOPeriod / Ts)  # number of lags for max freq.
Nend = round(minFOPeriod / Ts)  # number of lags for min freq.
# test with a cosine
y = np.cos(2 * np.pi * 300 * np.arange(0, 2 * Fs - 1) * Ts)  # 300 Hz, duration 2 secs
# Plot y
plt.subplot(211)
plt.plot(Ts * np.arange(0, len(y)), y)
plt.xlabel("time(s)")
plt.ylabel("Signal y(t)")

# ACF with max lag Nend
lags = plt.acorr(y, normed=False, maxlags=Nend)[0]  # Returning Lags
R = plt.acorr(y, normed=False, maxlags=Nend)[1] / len(y)  # returning R -> biased

firstIndex = np.where(lags == Nbegin)  # Find index of lag
Rpartial = R[firstIndex[0][0] :]  # Just the region of interest
Rmax = np.max(Rpartial)  # Max of Rpartial
relative_index_max = np.where(Rpartial == Rmax)
# Rpartial was just part of R, so recalculate the index
index_max = firstIndex + relative_index_max
lag_max = lags[index_max[0][0]]
# autocorrelation with normalized abscissa
plt.subplot(212)
plt.plot(lags * Ts, R)
plt.plot(lag_max * Ts, Rmax, "X")  # show the point
plt.xlabel("lag(s)")
plt.ylabel("Autocorrelation of y(t)")
plt.show()

F0 = 1 / (lag_max * Ts)  # estimated F0 frequency (Hz)
print(
    f"Rmax = {round(Rmax, 4)} lag_max= {round(lag_max, 4)} T= {round(lag_max*Ts, 4)}(s) Freq. = {F0} Hz\n"
)

# play freq. 3*F0
t = np.arange(0, 2, Ts)
aux = np.cos(2 * np.pi * 3 * F0 * t)
wr("Teste.wav", Fs, aux)  # Get the sample audio in the folder
